File: matlab/rtc_interface.m
classdef (ConstructOnLoad) rtc_interface < handle
    %RTC_INTERFACE Interface to the real-time control device. As far as
    %  possible this object is designed to be state-less. (The exception
    %  being the parameter list.)

    properties (Hidden)
        MAX_PACKET_SIZE = 64;
        CMD_GET_PAR_NAMES = 10;
        CMD_GET_PAR_SIZES = 11;
        CMD_GET_PAR_TYPES = 12;
        CMD_GET_PAR_VALUE = 20;
        CMD_SET_PAR_VALUE = 25;
        CMD_GET_STREAM    = 30;
        CMD_SOFT_RESET    = 90;
        CMD_ACK           = uint32(4291668991);
        STREAM_STATE_INACTIVE = 0;
        STREAM_STATE_ACTIVE   = 1;
        STREAM_STATE_FINISHED = 2;
        dev;
        libusb_notfound;
        libusb_warnings;
        timeout;
        intf;
        ep_in;
        ep_out;
        par_info;
        par_idx;
    end

    properties
        par; % Parameters accessible to the user. 
        opt; % Default options for the rig.
    end
    
    methods (Hidden)
        function send_raw(obj, data)
            % SEND_RAW  Send raw data to the RTC device (internal).
            transferred = libpointer('int32Ptr', 0);
            dataptr = libpointer('uint8Ptr', data);
            r = calllib('libusb','libusb_bulk_transfer', obj.dev, obj.ep_out, dataptr, length(data), transferred, obj.timeout);
            if r ~= 0
                error('Error in bulk transfer send');
            end
            if transferred.value ~= length(data)
                error('Failed to write all the data');
            end
        end
        
        function send_cmd(obj, cmd, data)
            % SEND_CMD  Send a command to the RTC device (internal).
            cmd = typecast(uint16(cmd), 'uint8');
            datalen = typecast(uint32(length(data)), 'uint8');
            if isempty(data)
                packet = horzcat(cmd, datalen);
                obj.send_raw(packet);
            else
                crc = typecast(crc32(data), 'uint8');
                packet = horzcat(cmd, datalen, data, crc);
                obj.send_raw(packet);
            end
        end
        
        function output = recv_raw(obj, nbytes)
            % RECV_RAW  Receive raw data from the RTC device (internal).
            if mod(nbytes, obj.MAX_PACKET_SIZE)
                warning('rtc_interface:recv_raw', 'recv_raw should be used with a multiple of MAX_PACKET_SIZE bytes to avoid potential data corruption');
            end
            if nbytes > 104857600
                % Check that we aren't trying to read too much data
                error('Trying to receive an insane amount of data');
            end
            dataptr = libpointer('uint8Ptr', zeros(1, nbytes, 'uint8'));
            transferred = libpointer('int32Ptr', 0);
            r = calllib('libusb','libusb_bulk_transfer', obj.dev, obj.ep_in, dataptr, nbytes, transferred, obj.timeout);
            if r ~= 0
                error('Error in bulk transfer read');
            end
            output = dataptr.value(1:transferred.value);
        end
        
        function output = recv_cmd(obj)
            % RECV_CMD  Receive the results of a command from the RTC device (internal).
            data = obj.recv_raw(obj.MAX_PACKET_SIZE);
            if length(data) < 4
                error('No intelligible response from RTC device');
            end
            data_len = typecast(data(1:4), 'uint32');
            % Check for a simple acknowledgement response
            if data_len == obj.CMD_ACK
                output = 1;
                return;
            end
            % Check how much data is needed
            while data_len + 8 > length(data)
                data_needed = data_len + 8 - length(data);
                if mod(data_needed, obj.MAX_PACKET_SIZE) ~= 0
                    data_needed = data_needed + obj.MAX_PACKET_SIZE - mod(data_needed, obj.MAX_PACKET_SIZE);
                end
                new_data = obj.recv_raw(data_needed);
                if isempty(new_data)
                    error('Timed out while waiting for data');
                end
                data = horzcat(data, new_data); %#ok<AGROW>
            end
            % Process the data
            output = data(5:data_len+4);
            output_crc = typecast(data(data_len+5:data_len+8), 'uint32');
            crc = crc32(output);
            if crc ~= output_crc
                error('Error in data stream returned from RTC device - CRC failed');
            end            
        end
        
        function get_par_list(obj)
            % GET_PAR_LIST Get the parameters available to access on the RTC device (internal).
            
            % Types that we might have to deal with
            type_sizes = containers.Map();
            type_sizes('b') = 1; type_sizes('h') = 2; type_sizes('i') = 4;
            type_sizes('B') = 1; type_sizes('H') = 2; type_sizes('I') = 4; 
            type_sizes('c') = 1; type_sizes('f') = 4;
            type_conv = containers.Map();
            type_conv('b') = 'int8'; type_conv('h') = 'int16'; type_conv('i') = 'int32';
            type_conv('B') = 'uint8'; type_conv('H') = 'uint16'; type_conv('I') = 'uint32';
            type_conv('f') = 'single'; type_conv('c') = 'char';
            % Get the parameter names
            obj.send_cmd(obj.CMD_GET_PAR_NAMES, []);
            names = char(obj.recv_cmd());
            % Get the parameter sizes
            obj.send_cmd(obj.CMD_GET_PAR_SIZES, []);
            sizes = typecast(obj.recv_cmd(), 'uint32');
            % Get the parameter types
            obj.send_cmd(obj.CMD_GET_PAR_TYPES, []);
            types = char(obj.recv_cmd());
            % Find the delimiters between names
            idx = [0, find(names == 0)];
            % Construct an appropriate structures to store all the information
            obj.par_info = struct('id', cell(1, length(idx) - 1), ...
                                  'name', '', ...
                                  'size', 0, ...
                                  'type', '', ...
                                  'count', 0);
            obj.par_idx = containers.Map();
            % Store all the information
            for i = 1:length(idx)-1
                obj.par_info(i).id = uint32(i - 1);
                obj.par_info(i).name = names(idx(i)+1:idx(i+1)-1);
                obj.par_info(i).size = sizes(i);
                obj.par_info(i).type = type_conv(types(i));
                obj.par_info(i).count = sizes(i)/type_sizes(types(i));
                obj.par_idx(obj.par_info(i).name) = i;
            end
        end
    end
    
    methods
        function obj = rtc_interface()
            %RTC_INTERFACE Interface to the real-time control device. As far as
            %  possible this object is designed to be state-less. (The exception
            %  being the parameter list.)
            
            % Load the library
            if ~libisloaded('libusb')
                [obj.libusb_notfound, obj.libusb_warnings] = loadlibrary('libusb','libusb.h');
            end
            calllib('libusb', 'libusb_init', libpointer);
            obj.dev = calllib('libusb', 'libusb_open_device_with_vid_pid', libpointer, hex2dec('0123'), hex2dec('4567'));
            if obj.dev.isNull()
                error('Unable to open the RTC device');
            else
                fprintf('Connected to RTC device\n');
            end
            r = calllib('libusb','libusb_claim_interface', obj.dev, 0);
            if r < 0
                error('Failed to claim the RTC device');
            end
            
            % Set the end points and interface of interest
            obj.intf = 0;
            obj.ep_out = 1;
            obj.ep_in = 129;

            % The timeout for any USB communications
            obj.timeout = 5000;

            % Get the parameter names
            obj.get_par_list();

            % A helper class for getting/setting parameters
            obj.par = rtc_parameters(obj);
            
            % Set the underlying device name
            obj.opt.device = 'BBB-RTC';
        end
        
        function delete(obj)
            % DELETE  Destroy the interface to the real-time controller.
            if ~obj.dev.isNull()
                calllib('libusb', 'libusb_close', obj.dev);
                fprintf('Released RTC device\n');
            end
            calllib('libusb', 'libusb_exit', libpointer);
        end
        
        function set_par(obj, names, values)
            % SET_PAR  Set the values of the specified parameters.
            %
            % OBJ.SET_PAR(NAME, VALUE) sets the value of the parameter NAME to VALUE.
            % Both NAME and VALUE can be cell arrays in the case of setting multiple
            % parameter values simultaneously.

            if ~iscell(names)
                names = {names};
                values = {values};
            end
            % Pack the data appropriately
            payload = uint8([]);
            for i = 1:length(names)
                idx = obj.par_idx(names{i});
                par_id = obj.par_info(idx).id;
                par_type = obj.par_info(idx).type;
                par_count = obj.par_info(idx).count;
                if length(values{i}) ~= par_count
                    error('Incorrect number of values supplied for parameter %s', names{i});
                end
                value = cast(values{i}, par_type);
                if strcmp(par_type, 'char')
                    payload = horzcat(payload, typecast(par_id, 'uint8'), ...
                                      cast(value, 'uint8')); %#ok<AGROW>
                else
                    payload = horzcat(payload, typecast(par_id, 'uint8'), ...
                                      typecast(value, 'uint8')); %#ok<AGROW>
                end
            end
            % Send the command
            obj.send_cmd(obj.CMD_SET_PAR_VALUE, payload);
            % Get the response
            if obj.recv_cmd() ~= 1
                error('Failed to set parameter values');
            end
        end
        
        function values = get_par(obj, names)
            % GET_PAR  Get the values of the specified parameters.
            %
            % OBJ.GET_PAR(NAME) gets the value of the parameter NAME. NAME can be a
            % cell array to get multiple parameter values simultaneously.
            
            if ~iscell(names)
                names = {names};
                islist = 0;
            else
                islist = 1;
            end
            % Get the ids associated with the names provided
            par_ids = zeros(1, length(names), 'uint32');
            par_idxs = zeros(1, length(names));
            for i = 1:length(names)
                par_idxs(i) = obj.par_idx(names{i});
                par_ids(i) = obj.par_info(par_idxs(i)).id;
            end
            % Request the parameter values
            obj.send_cmd(obj.CMD_GET_PAR_VALUE, typecast(par_ids, 'uint8'));
            data = obj.recv_cmd();
            % Reformat the data as appropriate
            idx = 1;
            data_list = cell(1, length(names));
            for i = 1:length(names)
                par_type = obj.par_info(par_idxs(i)).type;
                par_size = obj.par_info(par_idxs(i)).size;
                if strcmp(par_type, 'char')
                    data_list{i} = cast(data(idx:idx + par_size - 1), ...
                                        par_type);
                else
                    data_list{i} = typecast(data(idx:idx + par_size - 1), ...
                                            par_type);
                    if strcmp(par_type, 'single')
                        data_list{i} = cast(data_list{i}, 'double');
                    end
                end
                idx = idx + par_size;
            end
            % Return as a list or raw data depending on what was passed originally
            if islist
                values = data_list;
            else
                values = data_list{1};
            end
        end
        
        function set_stream(obj, stream, parameters, samples, downsample)
            % SET_STREAM  Set stream recording properties.
            %
            % OBJ.SET_STREAM(ID, NAMES, SAMPLES, DOWNSAMPLE) sets the stream with
            % identifier ID (where multiple streams are available) to record the
            % parameters given by the cell array NAMES. SAMPLES data points are
            % recorded and DOWNSAMPLE data points are discarded between each recorded
            % sample.
            %
            % Example
            %
            %   rtc.set_stream(0, {'x', 'out'}, 1000, 0);
            %
            % will set stream id 0 to record the parameters x and out. 1000 samples
            % will be returned with no data discarded.
            
            stream_name = sprintf('S%d', stream);
            if exist('samples', 'var') && ~isempty(samples)
                if samples > 0
                    obj.set_par([stream_name 'samples'], samples);
                end
            end
            if exist('downsample', 'var') && ~isempty(downsample)
                if downsample >= 0
                    obj.set_par([stream_name 'downsample'], downsample);
                end
            end
            if exist('parameters', 'var') && ~isempty(parameters)
                npars = obj.par_info(obj.par_idx([stream_name 'params'])).count;
                par_ids = zeros(1, npars, 'uint32');
                par_ids(:) = 4294967295;
                if length(parameters) > npars
                    error('Too many parameters provided');
                end
                for i = 1:length(parameters)
                    stream_par = obj.par_info(obj.par_idx(parameters{i}));
                    par_ids(i) = stream_par.id;
                    if ~(strcmp(stream_par.type, 'int32') || strcmp(stream_par.type, 'uint32') || strcmp(stream_par.type, 'single'))
                        error('Stream parameters must be either floats or ints (%s is %s)', stream_par.name, stream_par.type);
                    end
                end
                obj.set_par([stream_name 'params'], par_ids);
            end
        end
        
        function data = get_stream(obj, stream, return_struct)
            % GET_STREAM  Get the data from a particular stream.
            %
            % OBJ.GET_STREAM(ID) returns an array of data recorded in the stream given
            % by ID. If the stream is not ready, no data is returned.
            %
            % OBJ.GET_STREAM(ID, true) returns a structure with named fields containing
            % the data recorded in the stream given by ID.
            %
            % See also START_STREAM.
            stream_name = sprintf('S%d', stream);
            if obj.get_par([stream_name 'state']) ~= obj.STREAM_STATE_FINISHED
                data = [];
                return;
            end
            par_ids = obj.get_par([stream_name 'params']);
            par_count = find(par_ids >= length(obj.par_info), 1) - 1;
            par_ids = par_ids(1:par_count);
            obj.send_cmd(obj.CMD_GET_STREAM, typecast(uint32(stream), 'uint8'));
            raw_data = typecast(obj.recv_cmd(), 'uint32');
            raw_data = reshape(raw_data, [par_count, length(raw_data)/par_count]);
            if exist('return_struct', 'var') && return_struct
                for i = 1:par_count
                    par_type = obj.par_info(par_ids(i) + 1).type;
                    par_name = obj.par_info(par_ids(i) + 1).name;
                    data.(par_name) = typecast(raw_data(i, :), par_type);
                    if strcmp(par_type, 'single')
                        data.(par_name) = cast(data.(par_name), 'double');
                    end
                end
            else
                data = zeros(size(raw_data));
                for i = 1:par_count
                    par_type = obj.par_info(par_ids(i) + 1).type;
                    data(i, :) = cast(typecast(raw_data(i, :), par_type), 'double');
                end
            end
        end
        
        function result = start_stream(obj, stream)
            % START_STREAM  Start a stream recording.
            %
            % OBJ.START_STREAM(ID) starts the stream given by ID recording data with
            % the current parameters from SET_STREAM.
            %
            % See also SET_STREAM.
            stream_name = sprintf('S%d', stream);
            obj.set_par([stream_name 'state'], obj.STREAM_STATE_ACTIVE);
            if obj.get_par([stream_name 'state']) ~= obj.STREAM_STATE_INACTIVE
                result = 1;
            else
                result = 0;
            end
        end
        
        function data = run_stream(obj, stream, varargin)
            % RUN_STREAM  Start a stream recording and then return the captured data.
            %
            % OBJ.RUN_STREAM(ID) starts the stream given by ID and then returns the
            % captured data.
            %
            % OBJ.RUN_STREAM(ID, Name, Value) overrides the default options for running
            % the stream.
            %
            % Options
            %
            %     start: allowed values are true or false. Default true.
            %         Whether or not to start the stream running before waiting for
            %         available captured data.
            %
            %     wait_period: allowed values are a > 0. Default 0.1.
            %         The period of time the function should pause before checking if
            %         there is captured data available.
            %
            %     struct: allowed values are true or false. Default false.
            %         Whether or not to return the data as a structure.
            %
            % See also START_STREAM, GET_STREAM.
            p = inputParser();
            if ismethod(p, 'addParameter')
                % New versions of Matlab
                add_par = @p.addParameter;
            else
                % Old versions of Matlab
                add_par = @p.addParamValue;
            end
            add_par('start', true, @islogical);
            add_par('wait_period', 0.1, @(x)(x > 0));
            add_par('struct', false, @islogical);
            p.parse(varargin{:});
            stream_name = sprintf('S%d', stream);
            if p.Results.start
                if ~obj.start_stream(stream)
                    error('Failed to start stream - perhaps bad parameters');
                end
            elseif obj.get_par([stream_name 'state']) == obj.STREAM_STATE_INACTIVE
                error('Stream not already started');
            end
            while obj.get_par([stream_name 'state']) == obj.STREAM_STATE_ACTIVE
                pause(p.Results.wait_period);
            end
            data = obj.get_stream(stream, p.Results.struct);
        end
        
        function b = saveobj(~)
            b = [];
        end
    end
    
    methods (Static = true)
        function a = loadobj(~)
            a = [];
        end
    end
    
end

File: python/rtc_interface.py
# -*- coding: utf-8 -*-
"""An interface to the Beaglebone Black-based Real-Time Controller

Should work with both Python 2 and Python 3

Requires: numpy and libusb1 python modules and the appropriate libusb1
library (Windows or Linux)

@author: David A.W. Barton (david.barton@bristol.ac.uk)

Copyright (c) 2014, David A.W. Barton (david.barton@bristol.ac.uk)
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

1. Redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""


import usb1  # Import libusb1
import struct
import numpy as np
from array import array
import threading
from warnings import warn
from time import sleep

# Maximum USB packet size
MAX_PACKET_SIZE = 64

# Commands that can be sent
CMD_GET_PAR_NAMES = 10
CMD_GET_PAR_SIZES = 11
CMD_GET_PAR_TYPES = 12
CMD_GET_PAR_VALUE = 20
CMD_SET_PAR_VALUE = 25
CMD_GET_STREAM    = 30
CMD_SOFT_RESET    = 90

# Acknowledgement
CMD_ACK = struct.unpack("<I", b'\xFF\xAB\xCD\xFF')[0]

# Stream state
STREAM_STATE_INACTIVE = 0
STREAM_STATE_ACTIVE   = 1
STREAM_STATE_FINISHED = 2

# CRC set-up
CRC_INITIAL = 0xFFFFFFFF

# USB device lock
usblock = threading.Lock()


## rtc_exception - base class for RTC device errors
class rtc_exception(Exception):
    pass


## rtc_parameters - access parameters on the device
class rtc_parameters(object):
    """A simple class to simplify accesses to parameters"""

    _par_list = []

    def __init__(self, rt):
        self._rt = rt
        # To enable tab completion in IPython
        for name in rt.par_id:
            if not name in ["_rt", "_par_list"]:
                setattr(self, name, -1)
                self._par_list.append(name)

    def __getattribute__(self, name):
        super_self = super(rtc_parameters, self)
        if name in super_self.__getattribute__("_par_list"):
            return super_self.__getattribute__("_rt").get_par(name)
        else:
            return super_self.__getattribute__(name)

    def __setattr__(self, name, value):
        if name in self._par_list:
            return self._rt.set_par(name, value)
        else:
            return super(rtc_parameters, self).__setattr__(name, value)


## rtc_interface - main device interface
class rtc_interface(object):
    """RTC_INTERFACE Interface to the real-time control device. As far as
    possible this object is designed to be state-less. (The exception
    being the parameter list.)"""

    def __init__(self):
        """RTC_INTERFACE Interface to the real-time control device. As far as
        possible this object is designed to be state-less. (The exception
        being the parameter list.)"""

        # Set the end points and interface of interest
        self.intf = 0
        self.ep_out = 1
        self.ep_in = 129

        # Get the device
        self.context = usb1.LibUSBContext()
        self.dev = self.context.openByVendorIDAndProductID(0x0123, 0x4567)
        if self.dev is None:
            raise rtc_exception("RTC device not found")
        self.dev.claimInterface(self.intf)

        # The timeout for any USB communications
        self.timeout = 5000

        # Get the parameter names
        self.get_par_list()

        # A helper class for getting/setting parameters
        self.par = rtc_parameters(self)

    def send_raw(self, data):
        """SEND_RAW  Send raw data to the RTC device (internal)."""
        assert self.dev.bulkWrite(self.ep_out, data, self.timeout) == len(data)

    def send_cmd(self, cmd, data=None):
        """SEND_CMD  Send a command to the RTC device (internal)."""
        if (data is None) or (len(data) == 0):
            self.send_raw(struct.pack("<HI", cmd, 0))
        else:
            crc = crc_final(crc_update(CRC_INITIAL, data))
            self.send_raw(struct.pack("<HI%dsI" % len(data), cmd, len(data), data, crc))

    def recv_raw(self, nbytes):
        """RECV_RAW  Receive raw data from the RTC device (internal)."""
        if nbytes % MAX_PACKET_SIZE != 0:
            warn("recv_raw should be used with a multiple of MAX_PACKET_SIZE bytes to avoid potential data corruption")
        return self.dev.bulkRead(self.ep_in, nbytes, self.timeout)

    def recv_cmd(self):
        """RECV_CMD  Receive the results of a command from the RTC device (internal)."""
        data = self.recv_raw(MAX_PACKET_SIZE)
        if len(data) < 4:
            raise rtc_exception("No intelligible response from RTC device")
        else:
            # Check for a simple acknowledgement response
            if struct.unpack("<I", data)[0] == CMD_ACK:
                return True
            # Check how much data is needed
            data_len = struct.unpack("<I", data[0:4])[0]
            while data_len + 8 > len(data):
                # Need to get more data
                data_needed = data_len + 8 - len(data)
                if data_needed % MAX_PACKET_SIZE != 0:
                    data_needed += MAX_PACKET_SIZE - (data_needed % MAX_PACKET_SIZE)
                new_data = self.recv_raw(data_needed)
                if len(new_data) == 0:
                    raise rtc_exception("Timed out while waiting for data")
                data = data + new_data
            # Process the data
            output = data[4:data_len+4]
            output_crc = struct.unpack("<I", data[data_len+4:data_len+8])[0]
            crc = crc_final(crc_update(CRC_INITIAL, output))
            if crc != output_crc:
                raise rtc_exception("Error in data stream returned from RTC device - CRC failed")
            else:
                return output

    def release_usb(self):
        """RELEASE_USB Release the USB interface to the RTC device"""
        self.dev.releaseInterface(self.intf)

    def get_par_list(self):
        """GET_PAR_LIST Get the parameters available to access on the RTC device (internal)."""
        type_sizes = {'b' : 1, 'h' : 2, 'i' : 4, 'B' : 1, 'H' : 2, 'I' : 4, 'c' : 1, 'f' : 4}
        with usblock:
            self.send_cmd(CMD_GET_PAR_NAMES)
            names = self.recv_cmd()[0:-1].split(b"\0")
            self.send_cmd(CMD_GET_PAR_SIZES)
            sizes = struct.unpack("<%dI" % len(names), self.recv_cmd())
            self.send_cmd(CMD_GET_PAR_TYPES)
            types = struct.unpack("<%dc" % len(names), self.recv_cmd())
        self.par_info = []
        self.par_id = {}
        for i in range(0, len(names)):
            if type(names[i]) is str:
                par_name = names[i]
                par_type = types[i]
            else:
                par_name = names[i].decode()
                par_type = types[i].decode()
            par_info = {"id" : i,
                        "name" : par_name,
                        "type" : par_type,
                        "size" : sizes[i],
                        "count" : sizes[i]/type_sizes[par_type]}
            self.par_info.append(par_info)
            self.par_id[par_name] = i

    def set_par(self, names, values):
        """SET_PAR  Set the values of the specified parameters.
            
        OBJ.SET_PAR(NAME, VALUE) sets the value of the parameter NAME to VALUE.
        Both NAME and VALUE can be arrays in the case of setting multiple
        parameter values simultaneously."""
        if type(names) is str:
            names = [names]
            values = [values]
        # Pack the data appropriately
        payload = b''
        for (name, value) in zip(names, values):
            par_id = self.par_id[name]
            par_type = self.par_info[par_id]["type"]
            par_count = self.par_info[par_id]["count"]
            try:
                payload += struct.pack("<I%d%s" % (par_count, par_type), par_id, *value)
            except:
                payload += struct.pack("<I%d%s" % (par_count, par_type), par_id, value)
        # Send the command
        with usblock:
            self.send_cmd(CMD_SET_PAR_VALUE, payload)
            if self.recv_cmd():
                return True
            else:
                raise rtc_exception("Failed to set parameter values")

    def get_par(self, names):
        """GET_PAR  Get the values of the specified parameters.
            
        OBJ.GET_PAR(NAME) gets the value of the parameter NAME. NAME can be an
        array to get multiple parameter values simultaneously."""
        if type(names) is str:
            names = [names]
            islist = False
        else:
            islist = True
        # Get the ids associated with the names provided
        par_ids = []
        for name in names:
            assert(name in self.par_id)
            par_ids.append(self.par_id[name])
        # Request the parameter values
        with usblock:
            self.send_cmd(CMD_GET_PAR_VALUE, struct.pack("<%dI" % len(par_ids), *par_ids))
            data = self.recv_cmd()
        # Reformat the data as appropriate
        idx = 0
        data_list = []
        for par_id in par_ids:
            par_type = self.par_info[par_id]["type"]
            par_count = self.par_info[par_id]["count"]
            par_size = self.par_info[par_id]["size"]
            if par_type == 'c':
                data_list.append(data[idx:idx+par_size])
            elif par_count > 1:
                data_list.append(array(par_type,
                                       struct.unpack("<%d%s" % (par_count, par_type),
                                                     data[idx:idx+par_size])))
            else:
                data_list.append(struct.unpack("<%d%s" % (par_count, par_type),
                                               data[idx:idx+par_size])[0])
            idx += par_size
        # Return as a list or raw data depending on what was passed originally
        if islist:
            return data_list
        else:
            return data_list[0]

    def set_stream(self, stream, parameters=None, samples=None, downsample=None):
        """SET_STREAM  Set stream recording properties.
            
        OBJ.SET_STREAM(ID, NAMES, SAMPLES, DOWNSAMPLE) sets the stream with
        identifier ID (where multiple streams are available) to record the
        parameters given by the cell array NAMES. SAMPLES data points are
        recorded and DOWNSAMPLE data points are discarded between each recorded
        sample.
        
        Example
        
        rtc.set_stream(0, {'x', 'out'}, 1000, 0);
        
        will set stream id 0 to record the parameters x and out. 1000 samples
        will be returned with no data discarded."""
        stream_name = "S%d" % stream
        if samples is not None:
            assert(type(samples) is int)
            self.set_par(stream_name + "samples", samples)
        if downsample is not None:
            assert(type(downsample) is int)
            self.set_par(stream_name + "downsample", downsample)
        if parameters is not None:
            stream_par = stream_name + "params"
            npars = self.par_info[self.par_id[stream_par]]["count"]
            par_ids = []
            for name in parameters:
                par_id = self.par_id[name]
                par_ids.append(par_id)
                if not self.par_info[par_id]["type"] in ["i", "I", "f"]:
                    raise rtc_exception("Stream parameters must be either floats or ints")
            while len(par_ids) < npars:
                par_ids.append(0xFFFFFFFF)
            assert(len(par_ids) == npars)
            self.set_par(stream_par, par_ids)

    def get_stream(self, stream):
        """GET_STREAM  Get the data from a particular stream.
            
        OBJ.GET_STREAM(ID) returns an array of data recorded in the stream given
        by ID. If the stream is not ready, no data is returned.
        
        See also START_STREAM."""
        stream_name = "S%d" % stream
        if self.get_par(stream_name + "state") != STREAM_STATE_FINISHED:
            return None
        par_ids = self.get_par(stream_name + "params")
        par_count = 0
        for i in range(0, len(par_ids)):
            if par_ids[i] >= len(self.par_id):
                break
            par_count = i + 1
        par_ids = par_ids[0:par_count]
        par_types = "<"
        par_names = []
        data = {}
        for par_id in par_ids:
            par_type = self.par_info[par_id]["type"]
            par_types += par_type
            par_name = self.par_info[par_id]["name"]
            par_names.append(par_name)
            data[par_name] = array(par_type)
        self.send_cmd(CMD_GET_STREAM, struct.pack("<I", stream))
        raw_data = self.recv_cmd()
        idx = 0
        while idx < len(raw_data):
            tmp_data = struct.unpack(par_types, raw_data[idx:idx + par_count*4])
            for i in range(0, par_count):
                data[par_names[i]].append(tmp_data[i])
            idx += par_count*4
        return data

    def start_stream(self, stream):
        """START_STREAM  Start a stream recording.
            
        OBJ.START_STREAM(ID) starts the stream given by ID recording data with
        the current parameters from SET_STREAM.
            
        See also SET_STREAM."""
        stream_name = "S%d" % stream
        self.set_par(stream_name + "state", STREAM_STATE_ACTIVE)
        if self.get_par(stream_name + "state") != STREAM_STATE_INACTIVE:
            return True
        else:
            return False

    def run_stream(self, stream, start=True, wait_period=0.1):
        """RUN_STREAM  Start a stream recording and then return the captured data.
            
        OBJ.RUN_STREAM(ID) starts the stream given by ID and then returns the
        captured data.
            
        OBJ.RUN_STREAM(ID, Name, Value) overrides the default options for running
        the stream.
            
        Options
        
            start: allowed values are true or false. Default true.
                Whether or not to start the stream running before waiting for
                available captured data.
            
            wait_period: allowed values are a > 0. Default 0.1.
                The period of time the function should pause before checking if
                there is captured data available.
            
        See also START_STREAM, GET_STREAM."""
        stream_name = "S%d" % stream
        if start:
            if not self.start_stream(stream):
                raise rtc_exception("Failed to start stream - perhaps bad parameters")
        elif self.get_par(stream_name + "state") == STREAM_STATE_INACTIVE:
            raise rtc_exception("Stream not already started")
        while self.get_par(stream_name + "state") == STREAM_STATE_ACTIVE:
            sleep(wait_period)
        return self.get_stream(stream, struct)


## CRC tables - this is the same as used in the PNG standard
crc_table = np.zeros(256, dtype=np.uint32)

def crc_setup():
    """Set up the CRC tables"""
    for n in np.arange(0, 256, dtype=np.uint32):
        c = n
        for k in np.arange(0, 8, dtype=np.uint32):
            if c & 1:
                c = 0xEDB88320 ^ (c >> 1)
            else:
                c = c >> 1
        crc_table[n] = c

def crc_update(crc, data):
    """Update the CRC with the supplied data; use crc=CRC_INITIAL to start"""
    if type(data) is str:
        data = struct.unpack("<%dB" % len(data), data)
    for i in range(0, len(data)):
        crc = crc_table[(crc ^ data[i]) & 0xFF] ^ (crc >> 8)
    return crc

def crc_final(crc):
    """Finalise the CRC"""
    return crc ^ 0xFFFFFFFF

crc_setup()
